android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt

```diff
@@ -491,24 +491,52 @@ constructor(
     gatewayModeHeaderValue: String? = null,
     searchPath: String,
     patientRelatedResourceTypes: MutableList<ResourceType>,
+    currentPage: Int = 1,
   ) {
-    val resultBundle =
-      if (gatewayModeHeaderValue.isNullOrEmpty()) {
-        fhirResourceDataSource.getResource(searchPath)
-      } else
-        fhirResourceDataSource.getResourceWithGatewayModeHeader(
-          gatewayModeHeaderValue,
-          searchPath,
-        )
+    var currentPageUrl = "$searchPath&_page=$currentPage&_count=4"
+
+
+    // Fetch pages until there are no more next pages
+    while (currentPageUrl.isNotBlank()) {
+      val resultBundle = fetchResourceBundle(gatewayModeHeaderValue, currentPageUrl)
+      processResultBundleEntries(resultBundle.entry, patientRelatedResourceTypes)
+
+
+      // Extract the URL for the next page from the links
+      currentPageUrl = resultBundle.getLink("next").url ?: ""
+
+
+      // If next page URL is not available, it means we have fetched all pages
+      if (currentPageUrl.isBlank()) {
+        break
+      }
 
-    processResultBundleEntries(resultBundle, patientRelatedResourceTypes)
+      // Extracting page number from URL to increment it for the next page
+      val regex = Regex(".*&_page=(\\d+).*")
+      val matchResult = regex.matchEntire(currentPageUrl)
+      val nextPage = matchResult?.groupValues?.get(1)?.toIntOrNull()
+      nextPage?.let {
+        currentPageUrl = currentPageUrl.replace("_page=$nextPage", "_page=${nextPage + 1}")
```

Comment:
We already have the full url required for the next fetch so no need to parse the contents and extract the params. This way we can avoid a performance hit